### ISSUE TITLE
Do not reuse slave if retention time is set to 0

### ISSUE THREAD

**Issue by mvinkler** [jenkinsci/openstack-cloud-plugin#99]
Current behaviour: assume retention time is set to 0 in some Cloud Instance Template. If there are some jobs in the queue waiting for the particular instance, they will start executing there once the job currently running on the instance finished. After last job finishes, the instance is immediately destroyed.

Would it possible to force destroying instance after first job has finished (i.e. only one job will run on each created instance)?

**Comment by olivergondza** [comment]
There is some half-though (IMO) feature to do something alike in the plugin but I intend to reimplement this feature. What you suggest is on of the ways: admin will declare that machine from template is not reusable. An alternative would be for a job to decide, if the machine can be used or not. Which one do you prefer?

**Comment by zxiiro** [comment]
> Would it possible to force destroying instance after first job has finished (i.e. only one job will run on each created instance)?

Yes, configure the "single-shot instance" configuration at the job level. We apply this to all of our jobs and it causes OpenStack Cloud plugin to mark instances to be deleted at the end of a job.

**Comment by olivergondza** [comment]
@zxiiro, what I do not like about that implementation is, that every job needs to comply to the guideline that has in fact nothing to do with the build that has just completed. I am not aware of a use case when an author of a job needed to turn the node down once build is completed. What makes sense to me is:

- Admin declares a template its nodes will be single-use. Such nodes can have agents with elevated permissions so jobs can leave the machine in a state not suitable for further builds but all builds in the system are guaranteed to get such nodes clean (never reused before).
- Author of the job would like to make sure, the node was provisioned just for single build and never used before by any other build. No clutter of any kind accumulated.

Please let me know what your use-cases are.

**Comment by zxiiro** [comment]
@olivergondza Yes I would prefer allowing a template to be single-use at the admin level as well. I don't think this should be a job level permission except for the case where the job defines additional slaves to spin up. In which case I do think single-use there would make sense here because I would also like the possibility of the job itself also being allowed to define it's own node types and flavours.

**Comment by mvinkler** [comment]
>  admin will declare that machine from template is not reusable
@olivergondza for me this one is the exact solution I was hoping for

**Comment by warden** [comment]
It seems to be currently possible partially with DSL by defining the scope parameter:
```
openstackMachine: Cloud instances provisioning

Creates an Openstack machine, from an already globally defined cloud name and template without registering it as a Jenkins agent.

This step has a third optional String parameter named scope specifying how long is the server expected to live:
run
The booted machine will be kept alive whilst the current run is kicking. This is the default mode if no scope is defined.
unlimited
The booted machine will be kept alive forever. Mind the possible resources extarvation at the Openstack provider!
```

However, it is not clear to me how this goes in pair with the template's definition of the *Retention time* ? 
Maybe we should extend the scope to support _template retention policy_ value + additionally an integer value that would just destroy the instance after X minutes of inactivity time?

**Comment by olivergondza** [comment]
@warden server copes has nothing to do with slave retention time*. What is requested here is a valid feature that is not implemented yet.

*) Scope of an OpenStack server defines how long it is expected to run so it can be safely cleaned in case the plugin fails to delete it correctly or user decides to explicitly handle server lifespan. Note the `openstackMachine` provisions servers that are _not_ connected as Jenkins slaves.

**Comment by keysher** [comment]
HI, I have the same issue, what is the current status?

I have a lot of multibranch jobs with a label. The label matches with an Openstack Cloud template. But I want to use always a clean Openstack VM (No WS, maven cachés, etc...)

I use pipeline thru Jenkinsfile. Is there any way to get this working?

Regards!!

**Comment by codylample** [comment]
+1

**Comment by elyezer** [comment]
This would be very useful. I started using Jenkins Pipeline and I am running things in parallel and each one in one slave. Sometimes I have to run/install some stuff and having the slaves to be just dropped after a single usage is something interesting because some times the slave could be left in a broken state.

**Comment by elyezer** [comment]
@olivergondza shared that this PR https://github.com/jenkinsci/openstack-cloud-plugin/pull/209. It should get this issue fixed. I will take a look and state if that is really working and we can just close this issue.

**Comment by ajkyffin** [comment]
While PR #209 mostly works, on some images it takes a minute or so for the agent to connect to Jenkins, during which time they get marked for termination. The check for `retentionTime < 0` in `JCloudsRetentionStrategy.doCheck` needs to be changed to `retentionTime <= 0` but that causes some tests to fail.

**Comment by pjdarton** [comment]
We've just hit this issue too.
FYI At present, while the online-help for the retention time is the kind of functionality that is desired, the actual code behavior doesn't work as documented.

I believe that there's a race condition [here](https://github.com/jenkinsci/openstack-cloud-plugin/blob/4c398bb6facf91f85f44daba09460d002b6112f2/src/main/java/jenkins/plugins/openstack/compute/JCloudsRetentionStrategy.java#L53) where, if the slave is not assigned a job immediately after it has connected, or if it's offline simply because it hasn't come online _yet_, it can be marked for termination before Jenkins has assigned it some work.
e.g. a JNLP slave won't have `c.isConnecting()` true (because `JNLPLauncher`'s `launch` method is a no-op and hence it returns quickly but before the slave is online) but `c.isIdle()` will be true (as it isn't doing anything yet) and `c.getOfflineCause()` will be `OfflineCause.LaunchFailed()` (because that's what `SlaveComputer._connect(boolean)` will set it to if the JNLP connection isn't established before `JNLPLauncher`'s `launch` method returns).
As far as I can tell, this won't happen with `SSHLauncher` because its `launch` method doesn't return until the slave has actually connected, so this race condition will only affect JNLP slaves.

i.e. It looks like the [retention check](https://github.com/jenkinsci/openstack-cloud-plugin/blob/4c398bb6facf91f85f44daba09460d002b6112f2/src/main/java/jenkins/plugins/openstack/compute/JCloudsRetentionStrategy.java#L60) needs to also have a clause that says "_if_ the retention time is zero _and_ this instance has never run a build _then_ keep it".


Personally, what I would suggest is that this code gets split into multiple retention strategies.
1. Keep forever
1. Single use
1. Keep until idle for too long
...each of which can have their own configuration fields and associated online help.

All options will need a configurable "expect the slave to come online within X minutes" setting after which a non-connected slave is disposed of (although we should allow users to specify <=0 to mean "wait forever").
Both the single-use and keep-until-idle will need an idle retention time, because it's possible for a build to be cancelled before it's started running, so we could have a single-use VM started for a build that never happens, and we want to quietly dispose of the inactive VM after a while.
It'd be nice if the "forever" "until-idle" options had a "maximum duration" too, after which they won't be available for new work.

...and, to be honest, all of these options should really be put into a commonly-available plugin that can be shared between multiple cloud plugins, because the vSphere and docker (and probably AWS & Azure) plugins could all benefit from this sort of functionality.  It's crazy that every cloud plugin has to write its own.

**Comment by otrejoso** [comment]
Is there any workaround to avoid or reduce hitting this issue as much as possible? We started to hit this pretty much everyday.

**Comment by codylample** [comment]
A general workaround I've implemented with our pipeline jobs is to call `Jenkins.instanceOrNull.getSlave(env.NODE_NAME).terminate()` in an always/finally block at the end of the job. From what I've seen, it cleans up the instance on Jenkins and OpenStack and doesn't seem to cause job failures.

Note: Since it is a system call, it will require script approval unless implemented in a pipeline library (where I've chosen to use it).

**Comment by olivergondza** [comment]
There are several people reporting 0 retention time might not work reliably. This needs to be looked at.

**Comment by elyezer** [comment]
My use case is the single use one. I run the following groovy post-build:

```groovy
manager.build.getBuiltOn().toComputer().deleteSlave()
```

Note that I am not using the Jenkins Pipelines when doing the above. On the pipeline jobs I don't call that and I have not observed any issue. I will take a closer look to see if the offline node issue is happening.

**Comment by olivergondza** [comment]
Fix merged as #250 and #251.
